### PR TITLE
Add screenshots on failure for mobile apps

### PR DESCRIPTION
**Pull request by uarlouski**

### CODE REVIEW

File: vividus-plugin-web-app/src/main/java/org/vividus/selenium/screenshot/ScreenshotTaker.java [hunk 1]
```diff
@@ -154,14 +154,10 @@ public void takeScreenshot(Path screenshotFilePath, boolean viewportScreenshot)
...
-            Screenshot screenshot = new Screenshot();
-            screenshot.setData(screenshotData);
-            screenshot.setFileName(generateScreenshotFileName(screenshotName));
-            return Optional.of(screenshot);
-        }
-        return Optional.empty();
+        Screenshot screenshot = screenshotData.length > 0
+                ? new Screenshot(generateScreenshotFileName(screenshotName), screenshotData)
+                : null;
+        return Optional.ofNullable(screenshot);
```
Comment (valfirst) [inline vividus-plugin-web-app/src/main/java/org/vividus/selenium/screenshot/ScreenshotTaker.java:160]:
```java
return screenshotData.length > 0
            ? Optional.of(new Screenshot(generateScreenshotFileName(screenshotName), screenshotData))
            : Optional.empty();
```

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-plugin-web-app/src/main/java/org/vividus/selenium/screenshot/ScreenshotTaker.java (hunk 1 above)
Comment (uarlouski) [reply vividus-plugin-web-app/src/main/java/org/vividus/selenium/screenshot/ScreenshotTaker.java]: fixed, but imho too many method invocations per line

Comment (valfirst) [reply vividus-plugin-web-app/src/main/java/org/vividus/selenium/screenshot/ScreenshotTaker.java]: in this case you can keep old `if`

File: vividus-extension-selenium/src/main/java/org/vividus/bdd/monitor/AbstractScreenshotOnFailureMonitor.java [hunk 2]
```diff
@@ -37,26 +32,30 @@
...
-    @Inject private IWebDriverProvider webDriverProvider;
-    @Inject private IUiContext uiContext;
-    @Inject private IScreenshotTaker screenshotTaker;
+    public AbstractScreenshotOnFailureMonitor(EventBus eventBus, IBddRunContext bddRunContext,
+            IWebDriverProvider webDriverProvider)
+    {
+        this.eventBus = eventBus;
+        this.bddRunContext = bddRunContext;
+        this.webDriverProvider = webDriverProvider;
+        this.eventBus.register(this);
```
Comment (valfirst) [inline vividus-extension-selenium/src/main/java/org/vividus/bdd/monitor/AbstractScreenshotOnFailureMonitor.java:57]: why is it needed?

Comment (uarlouski) [reply vividus-extension-selenium/src/main/java/org/vividus/bdd/monitor/AbstractScreenshotOnFailureMonitor.java:57]: SubscriberRegisteringBeanPostProcessor doesn't support finding of subscribers in parent classes, this is should be separate commit

Comment (valfirst) [reply vividus-extension-selenium/src/main/java/org/vividus/bdd/monitor/AbstractScreenshotOnFailureMonitor.java:57]: let's fix this issue first?

Comment (uarlouski) [reply vividus-extension-selenium/src/main/java/org/vividus/bdd/monitor/AbstractScreenshotOnFailureMonitor.java:57]: in progress

Comment (uarlouski) [reply vividus-extension-selenium/src/main/java/org/vividus/bdd/monitor/AbstractScreenshotOnFailureMonitor.java:57]: Will rebase after https://github.com/vividus-framework/vividus/pull/961

**Review by valfirst — APPROVED** [review]

**Review by valfirst — APPROVED** [review]

**Pull request merged**